llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -27420,10 +27420,8 @@ SDValue DAGCombiner::XformToShuffleWithZero(SDNode *N) {
       }
 
       APInt Bits;
-      if (auto *Cst = dyn_cast<ConstantSDNode>(Elt))
-        Bits = Cst->getAPIntValue();
-      else if (auto *CstFP = dyn_cast<ConstantFPSDNode>(Elt))
-        Bits = CstFP->getValueAPF().bitcastToAPInt();
+      if (auto OptBits = Elt->bitcastToAPInt())
+        Bits = *std::move(OptBits);
       else
         return SDValue();
```

Comment:
Avoid the move and should invert this to early return on failure

Reply:
Do you mean:

```cpp
auto OptBits = Elt->bitcastToAPInt();
if (!OptBits)
  return SDValue();
APInt Bits = *OptBits;
```
?

Reply:
Yes. I'd do

```
  std::optional<APInt> Bits = Elt->bitcastToAPInt();
  if (!Bits)
    return SDValue();
  // Use *Bits

Reply:
Done